cranelift/codegen/src/ir/pcc.rs

```diff
@@ -257,135 +265,128 @@ impl BaseExpr {
-    fn min(lhs: &Expr, rhs: &Expr) -> Expr {
-        if lhs.base == BaseExpr::None && lhs.offset == 0 {
-            lhs.clone()
-        } else if rhs.base == BaseExpr::None && rhs.offset == 0 {
-            rhs.clone()
-        } else {
-            Expr {
-                base: BaseExpr::min(&lhs.base, &rhs.base),
-                offset: std::cmp::min(lhs.offset, rhs.offset),
-            }
-        }
-    }
-
-    /// Generalization of `max`: compute some Expr that is greater
-    /// than or equal to both inputs.
-    fn max(lhs: &Expr, rhs: &Expr) -> Expr {
-        if lhs.base == BaseExpr::None && lhs.offset == 0 {
-            rhs.clone()
-        } else if rhs.base == BaseExpr::None && rhs.offset == 0 {
-            lhs.clone()
-        } else {
-            Expr {
-                base: BaseExpr::max(&lhs.base, &rhs.base),
-                offset: std::cmp::max(lhs.offset, rhs.offset),
-            }
-        }
+        };
+        trace!("Expr::le: {lhs:?} {rhs:?} -> {result}");
+        result
     }
 
     /// Add one expression to another.
-    fn add(lhs: &Expr, rhs: &Expr) -> Option<Expr> {
-        if lhs.base == rhs.base {
-            Some(Expr {
+    fn add(lhs: &Expr, rhs: &Expr) -> Expr {
+        let Some(offset) = lhs.offset.checked_add(rhs.offset) else {
+            return Expr::max_value();
+        };
+        let result = if lhs.base == rhs.base {
+            Expr {
                 base: lhs.base.clone(),
-                offset: lhs.offset.checked_add(rhs.offset)?,
-            })
+                offset,
+            }
         } else if lhs.base == BaseExpr::None {
-            Some(Expr {
+            Expr {
                 base: rhs.base.clone(),
-                offset: lhs.offset.checked_add(rhs.offset)?,
-            })
+                offset,
+            }
         } else if rhs.base == BaseExpr::None {
-            Some(Expr {
+            Expr {
                 base: lhs.base.clone(),
-                offset: lhs.offset.checked_add(rhs.offset)?,
-            })
+                offset,
+            }
         } else {
-            Some(Expr {
+            Expr {
                 base: BaseExpr::Max,
                 offset: 0,
-            })
-        }
+            }
+        };
+        trace!("Expr::add: {lhs:?} + {rhs:?} -> {result:?}");
+        result
     }
 
     /// Add a static offset to an expression.
     pub fn offset(lhs: &Expr, rhs: i64) -> Option<Expr> {
-        let offset = lhs.offset.checked_add(rhs)?;
+        let offset = lhs.offset.checked_add(rhs.into())?;
         Some(Expr {
             base: lhs.base.clone(),
             offset,
         })
     }
 
+    /// Determine if we can know the difference between two expressions.
+    pub fn difference(lhs: &Expr, rhs: &Expr) -> Option<i64> {
+        match (lhs.base, rhs.base) {
+            (BaseExpr::Max, _) | (_, BaseExpr::Max) => None,
+            (a, b) if a == b => i64::try_from(lhs.offset.checked_sub(rhs.offset)?).ok(),
```

Comment:
It seems like `Max` should be a variant of `Expr` rather than `BaseExpr`? Because it seems like offsets shouldn't matter once we reach `Max`? So having both a non-zero offset and a base of `Max` seems like something we should make unrepresentable.

Reply:
Good idea, fixed!